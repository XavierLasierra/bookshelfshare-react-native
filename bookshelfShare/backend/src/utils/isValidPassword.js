const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
