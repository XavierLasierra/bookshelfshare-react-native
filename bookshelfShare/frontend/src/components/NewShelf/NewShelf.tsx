import React, { useState } from 'react';
import {
  Text, TouchableOpacity, View, TextInput
} from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
// import { useDispatch, useSelector } from 'react-redux';
import Header from '../Header/Header';

import ShelfIcon from '../../assets/shelfIcon.svg';
import styles from './newShelf.styles';
import globalStyles from '../../styles/global.styles';

interface Props {
  navigation: any,
}

export default function NewShelf({ navigation }: Props) {
  // const dispatch = useDispatch();
  // const { token, refreshToken } = useSelector((store: any) => store.tokens);
  const [name, setName] = useState('');
  const [rows, setRows] = useState('');
  const [columns, setColumns] = useState('');

  function handleShelfCreation() {
    navigation.push();
  }

  function handleNameChange(text: string) {
    setName(text);
  }

  function handleRowsChange(text: string) {
    setRows(text);
  }

  function handleColumnsChange(text: string) {
    setColumns(text);
  }

  return (
    <SafeAreaView style={globalStyles.mainContainer}>
      <Header Logo={ShelfIcon} BackButton navigation={navigation} />
      <View style={styles.createShelfContainer}>
        <Text style={globalStyles.title}>
          New shelf
        </Text>
        <View style={styles.formContainer}>
          <View style={globalStyles.inputContainer}>
            <Text style={globalStyles.inputLabel}>name</Text>
            <TextInput
              style={globalStyles.input}
              onChangeText={handleNameChange}
              value={name}
              testID="nameInput"
              maxLength={25}
            />
          </View>
          <View style={styles.shelfSizeContainer}>
            <View style={[globalStyles.inputContainer, styles.smallInputContainer]}>
              <Text style={globalStyles.inputLabel}>rows</Text>
              <TextInput
                style={globalStyles.input}
                onChangeText={handleRowsChange}
                value={rows}
                testID="rowsInput"
                keyboardType="number-pad"
                maxLength={2}
              />
            </View>
            <View style={[globalStyles.inputContainer, styles.smallInputContainer]}>
              <Text style={globalStyles.inputLabel}>columns</Text>
              <TextInput
                style={globalStyles.input}
                onChangeText={handleColumnsChange}
                value={columns}
                testID="columnsInput"
                keyboardType="number-pad"
                maxLength={2}
              />
            </View>
          </View>
          <TouchableOpacity
            style={[globalStyles.button, styles.createButton]}
            onPress={handleShelfCreation}
            testID="createButton"
          >
            <Text style={globalStyles.buttonText}>Create</Text>
          </TouchableOpacity>
        </View>
      </View>
    </SafeAreaView>
  );
}
