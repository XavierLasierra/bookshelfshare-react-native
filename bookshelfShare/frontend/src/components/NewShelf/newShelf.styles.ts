import { StyleSheet } from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  createShelfContainer: {
    padding: stylesConstants.padding.normal,
    alignItems: 'center'
  },
  formContainer: {
    width: '100%',
    height: 300,
    justifyContent: 'space-between'
  },
  createButton: {
    marginTop: 20
  },
  shelfSizeContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  smallInputContainer: {
    width: '45%'
  }
});
