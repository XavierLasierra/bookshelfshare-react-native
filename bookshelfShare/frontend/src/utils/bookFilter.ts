export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
