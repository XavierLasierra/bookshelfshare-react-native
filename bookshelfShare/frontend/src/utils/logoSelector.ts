import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
