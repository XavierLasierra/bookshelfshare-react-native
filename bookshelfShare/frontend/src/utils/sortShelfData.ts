export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
