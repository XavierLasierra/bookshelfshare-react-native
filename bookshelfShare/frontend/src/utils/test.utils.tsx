import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
