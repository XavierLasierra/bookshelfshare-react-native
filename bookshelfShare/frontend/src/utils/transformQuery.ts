interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
