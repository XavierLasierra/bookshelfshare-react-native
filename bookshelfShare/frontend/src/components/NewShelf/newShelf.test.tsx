import React from 'react';
import NewShelf from './NewShelf';
import { render, fireEvent } from '../../utils/test.utils';
import { createShelf } from '../../redux/actions/userShelves.creator';
import userShelvesActions from '../../redux/actions/userShelves.actions';

jest.mock('../../redux/actions/userShelves.creator');

const navigation = {
  pop: jest.fn()
};

const route = {
  params: {
    loggedUserId: '1'
  }
};

describe('Given a NewShelf component', () => {
  describe('When it is rendered', () => {
    let screen: any;
    beforeEach(() => {
      (createShelf as jest.Mock).mockReturnValue({ type: '' });
      screen = render(
        <NewShelf navigation={navigation} route={route} />
      );
    });
    test('Then should match the snapshot', () => {
      expect(screen).toMatchSnapshot();
    });

    describe('And createButton is clicked with an empty name', () => {
      beforeEach(() => {
        const createButton = screen.getByTestId('createButton');
        fireEvent.press(createButton);
      });
      test('Then should render a text with testId invalidShelfName', () => {
        expect(screen.queryByTestId('invalidShelfName')).not.toBe(null);
      });

      describe('And you focus on nameInput', () => {
        test('Then should not render a text with testId invalidShelfName', () => {
          const nameInput = screen.getByTestId('nameInput');
          fireEvent(nameInput, 'focus');

          expect(screen.queryByTestId('invalidShelfName')).toBe(null);
        });
      });
    });

    describe('And you type on nameInput', () => {
      beforeEach(() => {
        const nameInput = screen.getByTestId('nameInput');
        fireEvent(nameInput, 'focus');
        fireEvent.changeText(nameInput, 'name');
      });
      describe('And createButton is clicked', () => {
        beforeEach(() => {
          const createButton = screen.getByTestId('createButton');
          fireEvent.press(createButton);
        });
        test('Then should render a text with testId invalidShelfSize', () => {
          expect(screen.queryByTestId('invalidShelfSize')).not.toBe(null);
        });

        describe('And you focus on rowsInput', () => {
          test('Then should not render a text with testId invalidShelfSize', () => {
            const rowsInput = screen.getByTestId('rowsInput');
            fireEvent(rowsInput, 'focus');

            expect(screen.queryByTestId('invalidShelfSize')).toBe(null);
          });
        });

        describe('And you focus on columnsInput', () => {
          test('Then should not render a text with testId invalidShelfSize', () => {
            const columnsInput = screen.getByTestId('columnsInput');
            fireEvent(columnsInput, 'focus');

            expect(screen.queryByTestId('invalidShelfSize')).toBe(null);
          });
        });
      });
    });

    describe('And you type on nameInput and rowsInput', () => {
      beforeEach(() => {
        const nameInput = screen.getByTestId('nameInput');
        fireEvent(nameInput, 'focus');
        fireEvent.changeText(nameInput, 'name');
        const rowsInput = screen.getByTestId('rowsInput');
        fireEvent(rowsInput, 'focus');
        fireEvent.changeText(rowsInput, '2');
      });
      describe('And createButton is clicked', () => {
        beforeEach(() => {
          const createButton = screen.getByTestId('createButton');
          fireEvent.press(createButton);
        });
        test('Then should render a text with testId invalidShelfSize', () => {
          expect(screen.queryByTestId('invalidShelfSize')).not.toBe(null);
        });
      });
    });

    describe('And you type on nameInput, rowsInput and columnsInput', () => {
      describe('And createButton is clicked', () => {
        describe('And shelves update is loading', () => {
          beforeEach(() => {
            const nameInput = screen.getByTestId('nameInput');
            fireEvent(nameInput, 'focus');
            fireEvent.changeText(nameInput, 'name');
            const rowsInput = screen.getByTestId('rowsInput');
            fireEvent.changeText(rowsInput, '2');
            const columnsInput = screen.getByTestId('columnsInput');
            fireEvent.changeText(columnsInput, '3');

            const createButton = screen.getByTestId('createButton');
            fireEvent.press(createButton);
          });

          test('Then should call createShelf with the first parameter containing shelvesInfomation', () => {
            expect((createShelf as jest.Mock).mock.calls[0][0]).toEqual({ name: 'name', users: ['1'], shelf: [3, 3] });
          });

          test('Then should render an ActivityIndicator with testId activityIndicator', () => {
            expect(screen.queryByTestId('activityIndicator')).not.toBe(null);
          });
        });
        describe('And shelves update is resolved', () => {
          test('Then navigation.pop should have been called', () => {
            (createShelf as jest.Mock).mockReturnValue({ type: userShelvesActions.ADD_NEW_SHELF });
            const newScreen = render(
              <NewShelf navigation={navigation} route={route} />
            );

            const nameInput = newScreen.getByTestId('nameInput');
            fireEvent(nameInput, 'focus');
            fireEvent.changeText(nameInput, 'name');
            const rowsInput = newScreen.getByTestId('rowsInput');
            fireEvent.changeText(rowsInput, '2');
            const columnsInput = newScreen.getByTestId('columnsInput');
            fireEvent.changeText(columnsInput, '3');

            const createButton = newScreen.getByTestId('createButton');
            fireEvent.press(createButton);

            expect(navigation.pop).toHaveBeenCalled();
          });
        });
      });
    });
  });
});
