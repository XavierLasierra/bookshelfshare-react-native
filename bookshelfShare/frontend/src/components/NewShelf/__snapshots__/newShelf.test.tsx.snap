// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a NewShelf component When it is rendered Then should match the snapshot 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#F5F5F5",
      "minHeight": "100%",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 20,
        "opacity": 1,
        "position": "absolute",
        "top": 20,
        "zIndex": 1,
      }
    }
    testID="backButton"
  >
    <IconMock
      height={30}
      style={
        Object {
          "transform": Array [
            Object {
              "rotateZ": "180deg",
            },
          ],
        }
      }
      width={30}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "#F5F5F5",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#D9D8C4",
          "borderBottomLeftRadius": 25,
          "justifyContent": "center",
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <IconMock
        height={45}
        width={45}
      />
    </View>
    <View
      style={
        Object {
          "height": 0,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#D9D8C4",
            "height": 20,
            "width": 20,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F5F5F5",
              "borderTopRightRadius": 25,
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "padding": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3E3C37",
          "fontSize": 35,
          "fontWeight": "500",
        }
      }
    >
      New shelf
    </Text>
    <View
      style={
        Object {
          "height": 300,
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "height": 70,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 12,
              "fontWeight": "bold",
              "transform": Array [
                Object {
                  "translateX": 5,
                },
                Object {
                  "translateY": 20,
                },
              ],
              "zIndex": 1,
            }
          }
        >
          name
        </Text>
        <TextInput
          maxLength={25}
          onChangeText={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#ECECEC",
              "borderRadius": 10,
              "borderWidth": 1,
              "fontSize": 20,
              "height": "100%",
              "paddingHorizontal": 30,
              "paddingTop": 30,
            }
          }
          testID="nameInput"
          value=""
        />
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 70,
                "width": "100%",
              },
              Object {
                "width": "45%",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontSize": 12,
                "fontWeight": "bold",
                "transform": Array [
                  Object {
                    "translateX": 5,
                  },
                  Object {
                    "translateY": 20,
                  },
                ],
                "zIndex": 1,
              }
            }
          >
            rows
          </Text>
          <TextInput
            keyboardType="number-pad"
            maxLength={1}
            onChangeText={[Function]}
            onFocus={[Function]}
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderColor": "#ECECEC",
                "borderRadius": 10,
                "borderWidth": 1,
                "fontSize": 20,
                "height": "100%",
                "paddingHorizontal": 30,
                "paddingTop": 30,
              }
            }
            testID="rowsInput"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "height": 70,
                "width": "100%",
              },
              Object {
                "width": "45%",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "fontSize": 12,
                "fontWeight": "bold",
                "transform": Array [
                  Object {
                    "translateX": 5,
                  },
                  Object {
                    "translateY": 20,
                  },
                ],
                "zIndex": 1,
              }
            }
          >
            columns
          </Text>
          <TextInput
            keyboardType="number-pad"
            maxLength={1}
            onChangeText={[Function]}
            onFocus={[Function]}
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "borderColor": "#ECECEC",
                "borderRadius": 10,
                "borderWidth": 1,
                "fontSize": 20,
                "height": "100%",
                "paddingHorizontal": 30,
                "paddingTop": 30,
              }
            }
            testID="columnsInput"
            value=""
          />
        </View>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#000000",
            "borderRadius": 10,
            "display": "flex",
            "height": 70,
            "justifyContent": "center",
            "marginTop": 20,
            "opacity": 1,
            "width": "100%",
          }
        }
        testID="createButton"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 20,
              "fontWeight": "500",
            }
          }
        >
          Create
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
